Source/Controllers/SubviewControllers/SPTablesList.m

```diff
@@ -1678,28 +1678,50 @@ - (void)tableView:(NSTableView *)aTableView setObjectValue:(id)anObject forTable
 		// Table has invalid name, and since we trimmed whitespace and checked for empty string, this means there is already a table with that name
 		[NSAlert createWarningAlertWithTitle:NSLocalizedString(@"Error", @"error") message:[NSString stringWithFormat: NSLocalizedString(@"The name '%@' is already used.", @"message when trying to rename a table/view/proc/etc to an already used name"), newTableName] callback:nil];
 		return;
-	}
-
-	@try {
-		// first: update the database
-		[self _renameTableOfType:selectedTableType from:selectedTableName to:newTableName];
-        
-        // second : unpin selectedTableName and pin newTableName
-        [self handlePinnedTableRenameFrom:selectedTableName To:newTableName];
-
-		// third: do full refresh
-        [self updateTables:self];
-        
-	}
-	@catch (NSException * myException) {
-		[NSAlert createWarningAlertWithTitle:NSLocalizedString(@"Error", @"error") message:[myException reason] callback:nil];
-	}
-
-	// Set window title to reflect the new table name
-	[tableDocumentInstance updateWindowTitle:self];
-
-	// Query the structure of all databases in the background (mainly for completion)
-	[[tableDocumentInstance databaseStructureRetrieval] queryDbStructureInBackgroundWithUserInfo:@{@"forceUpdate" : @YES}];
+  }
+  
+  __block BOOL isRenamed = NO;
+  if ([prefs boolForKey:SPQueryWarningEnabled]) {
+    [NSAlert createDefaultAlertWithTitle:NSLocalizedString(@"Rename table", @"Rename table")
+                                 message:[NSString stringWithFormat:NSLocalizedString(@"Do you want to rename '%@' table to '%@'?", @"rename table description"), 
+                                          selectedTableName,
+                                          newTableName]
+                      primaryButtonTitle:NSLocalizedString(@"Confirm", @"table structure : indexes : delete index : error 1553 : delete index and FK button")
+                    primaryButtonHandler:^{
+      [self renameTableOfType:self->selectedTableType from:self->selectedTableName to:newTableName];
+      isRenamed = YES;
```

Comment:
Want to confirm - is this callback logic executed asynchronously? If so, does the function return before the rename is completed with the window title update and Query the structure of all databases calls happening immediately, before the rename actually occurs? In other words, is the callback a deferred execution?

Reply:
I don't see how it could be async here? Could you give more details? A help on testing would really help because I don't really have knowledge about objective-c, just want to solve the problem of the tool.

To be honest, with the previous way I tried to use, it was ok to ensure blocking the logic when showing the pop-up. By this ne way, I don't really like it because of reusing the component but now had to duplicate the same logic.

Reply:
Sorry for the delay! It's been a busy week + COVID
@Kaspik Could you weigh in on how this handler works and confirm?

@vuon9 This might help - https://developer.apple.com/documentation/swift/calling-objective-c-apis-asynchronously
I believe the overall function is allowed to return _before_ a user interacts with the button to confirm or reject. Therefore, any code that you'd want to run IFF the user hits Confirm must be in the body of the buttonHandler and not in the general function body

Reply:
Hope you feel well, i will look at it again when I have time.

Reply:
I don't remember honestly, but as long as this works, sounds safe to me. Test both options on the alert :)

Reply:
Sounds good - @vuon9 have you tested both options on the alert?

Reply:
@vuon9 ?

Reply:
I didn't have time to test this yet. Let me ping you guys once again when it's ready. Thanks

Reply:
@vuon9 Did you get a chance to double check this PR?

Reply:
Not yet, sorry about that. I will try in this weekend.

Reply:
Hello,

https://github.com/user-attachments/assets/83ae72c2-7c3b-48ae-a0ab-eb589a01ad6c

With current implementation, I tested both Cancel/Confirm. The table list updated in sync after the table name changed. When cancelling, the table's name stays unchanged.

Reply:
Nice work!